### PR TITLE
feat: add prototype open sharing and add sharing stateful tests

### PR DESCRIPTION
**Pull request by ZhiHanZ**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

The PR adds:
1. documentation about share endpoint and share protocol
2. basic share protocol implementation
3. basic unit tests
4. s3 stateful tests about sharing.

Closes #issue

### CODE REVIEW

File: src/query/sharing-endpoint/src/handlers.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    let requester = credentials.token.clone();
+    let input = models::LambdaInput::new(
+        credentials.token.clone(),
+        share_name,
+        requester,
+        table_name,
+        request_files,
+        None,
+    );
+    println!("input: {:?}", input.clone());
```
Comment (lichuang) [inline src/query/sharing-endpoint/src/handlers.rs:41]: debug code？

File: tests/suites/1_stateful/06_sharing/06_0000_basic_sharing.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+echo "SHOW SHARES" | $SHARING_MYSQL_CLIENT_CONNECT | awk '{print $(NF-4), $(NF-3), $(NF-2), $(NF-1), $NF}'
+# bug: show database has no shared database but it still exists when create database
+echo "CREATE DATABASE if not exists shared FROM SHARE test_tenant.test_share" | $SHARING_MYSQL_CLIENT_CONNECT
+echo "SELECT * FROM shared.t1" | $SHARING_MYSQL_CLIENT_CONNECT
+
+
+## Drop table.
+echo "drop database if exists shared"  | $SHARING_MYSQL_CLIENT_CONNECT
+echo "drop share if exists test_share" | $MYSQL_CLIENT_CONNECT
+echo "drop database if exists test_database" | $MYSQL_CLIENT_CONNECT
```
Comment (lichuang) [inline tests/suites/1_stateful/06_sharing/06_0000_basic_sharing.sh:30]: can add some test that query data from other tenent via share table?

Comment (ZhiHanZ) [reply tests/suites/1_stateful/06_sharing/06_0000_basic_sharing.sh:30]: https://github.com/datafuselabs/databend/blob/968c6644ded42d5d81b734bd15b92a077515e031/tests/suites/1_stateful/06_sharing/06_0000_basic_sharing.result
here is the test

File: src/query/sharing-endpoint/src/accessor.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_base::base::Singleton;
```
Comment (Xuanwo) [inline src/query/sharing-endpoint/src/accessor.rs:17]: Singleton has been removed.

File: src/query/sharing-endpoint/src/accessor.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+use crate::configs::Config;
+use crate::models;
+use crate::models::PresignFileResponse;
+use crate::models::SharedTableResponse;
+
+static SHARING_ACCESSOR: OnceCell<Singleton<Arc<SharingAccessor>>> = OnceCell::new();
+
+#[derive(Clone)]
+pub struct SharingAccessor {
+    op: Arc<Operator>,
```
Comment (Xuanwo) [inline src/query/sharing-endpoint/src/accessor.rs:34]: Operator is cheap to clone and don't need `Arc`.

File: src/query/sharing-endpoint/src/accessor.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        if arr.len() > 2 {
+            return arr[2..].join("/");
+        }
+        loc.to_string()
+    };
+}
+
+impl SharingAccessor {
+    pub async fn init(cfg: &Config, v: Singleton<Arc<SharingAccessor>>) -> Result<()> {
+        let op = init_operator(&cfg.storage.params)?;
```
Comment (Xuanwo) [inline src/query/sharing-endpoint/src/accessor.rs:67]: Please adapt with the new `GlobalInstance` API

File: src/query/sharing-endpoint/src/services.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+use common_base::base::GlobalIORuntime;
+use common_base::base::Runtime;
+use common_base::base::SingletonImpl;
+use common_exception::Result;
+
+use crate::accessor::SharingAccessor;
+use crate::configs::Config;
+
+// hold singleton services.
+pub struct SharingServices {
```
Comment (Xuanwo) [inline src/query/sharing-endpoint/src/services.rs:27]: ditto. `GlobalInstace` will save us from `UnsafeCell` operations.

File: src/query/sharing-endpoint/src/services.rs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+// hold singleton services.
+pub struct SharingServices {}
+
+impl SharingServices {
+    pub async fn init(config: Config) -> Result<()> {
+        // init global instance singleton
+        GlobalInstance::init_production();
+
+        GlobalIORuntime::init(config.storage.num_cpus as usize)?;
```
Comment (Xuanwo) [inline src/query/sharing-endpoint/src/services.rs:30]: `GlobalIORuntime` should have been initiated.

**Review by Xuanwo — APPROVED** [review]
Mostly LGTM

**Review by lichuang — APPROVED** [review]

**Pull request merged**